### PR TITLE
Creating multiple policies in one tx

### PR DESCRIPTION
**Pull request by vzotova**
**Type of PR:**
- [X] Feature

**Required reviews:** 
- [X] 3

**What this does:**
One new function in `PolicyManager` to create multiple policies in one tx

**Issues fixed/closed:**
Refs #2508

**Why it's needed:**
Decrease gas cost when multiple policies needed with the same parameters

### CODE REVIEW

**Comment by KPrasch** [comment]
@vzotova - Do you think we need to run another EPIC-style branch since this will require a DAO proposal?

**Comment by KPrasch** [comment]
@vzotova - As discussed I've temporarily skipped the upgradeability test in PR #2644 and opened follow-up issue #2645 .  A rebase over main may resolve your CI failure for now.

File: tests/contracts/main/policy_manager/test_policy_manager.py [hunk 1]
```diff
@@ -517,6 +517,240 @@ def test_create_revoke(testerchain, escrow, policy_manager):
...
+    # Can't create policy using timestamp from the past
+    with pytest.raises((TransactionFailed, ValueError)):
+        tx = policy_manager.functions.createPolicies(policies, policy_sponsor, current_timestamp - 1, [node1])\
+            .transact({'from': policy_sponsor, 'value': 2 * value})
+        testerchain.wait_for_receipt(tx)
+
+    # Can't create two policies with the same id
+    with pytest.raises((TransactionFailed, ValueError)):
+        tx = policy_manager.functions.createPolicies([policy_id_1, policy_id_1], policy_sponsor, end_timestamp, [node1]) \
+            .transact({'from': policy_sponsor, 'value': 2 * value, 'gas_price': 0})
```
Comment (cygnusv) [inline tests/contracts/main/policy_manager/test_policy_manager.py:569]: Does `gas_price` work? AFAIK, it should be `gasPrice`

**Review by cygnusv — APPROVED** [review]

File: tests/contracts/main/policy_manager/test_policy_manager.py (hunk 1 above)
Comment (vzotova) [reply tests/contracts/main/policy_manager/test_policy_manager.py:569]: Checked again: yeah, `gas_price` works, and `gasPrice` too. `gasPrice` on web3 level transformed to `gas_price` for eth_tester level. I'd prefer leave as is (`gas_price`)

Comment (KPrasch) [reply tests/contracts/main/policy_manager/test_policy_manager.py:569]: The (good?) news is that the newest version of web3.py uses underscored names for all RPC endpoints.

**Review by derekpierre — APPROVED** [review]
🎸

File: nucypher/blockchain/eth/sol/source/contracts/PolicyManager.sol [hunk 2]
```diff
@@ -253,68 +253,161 @@ contract PolicyManager is Upgradeable {
...
+            _policyIds.length > 1
+        );
+
+        require(address(this).balance <= MAX_BALANCE);
+        uint16 currentPeriod = getCurrentPeriod();
+        uint16 endPeriod = uint16(_endTimestamp / secondsPerPeriod) + 1;
+        uint256 numberOfPeriods = endPeriod - currentPeriod;
+
+        uint128 feeRate = uint128(msg.value.div(_nodes.length) / numberOfPeriods / _policyIds.length);
+        require(feeRate > 0 && feeRate * numberOfPeriods * _nodes.length * _policyIds.length == msg.value);
```
[the remark this answers is not in the document]
Comment (vzotova) [reply nucypher/blockchain/eth/sol/source/contracts/PolicyManager.sol:306]: yep, samle expiration date, same Ursulas

File: nucypher/blockchain/eth/sol/source/contracts/PolicyManager.sol [hunk 3]
```diff
@@ -253,68 +253,161 @@ contract PolicyManager is Upgradeable {
...
+        address _policyOwner,
+        uint64 _endTimestamp,
+        address[] calldata _nodes
+    )
+        external payable
+    {
+        require(
+            _endTimestamp > block.timestamp &&
+            msg.value > 0 &&
+            _policyIds.length > 1
```
[the remark this answers is not in the document]
Comment (vzotova) [reply nucypher/blockchain/eth/sol/source/contracts/PolicyManager.sol:297]: it was a little bit easier to maintain and optimize right now, but it definitely can be possible to have only one method

**Review by KPrasch — COMMENTED** [review]
This PR is looking good, although I'm holding off on approval until I can work on top of it to implement the python API a bit, as that will result in a deeper review on my end. 👍🏻

**Pull request merged**